addons/portal/controllers/mail.py

```diff
@@ -221,9 +221,14 @@ def portal_message_fetch(self, res_model, res_id, domain=False, limit=10, offset
             if not request.env['res.users'].has_group('base.group_user'):
                 domain = expression.AND([Message._get_search_domain_share(), domain])
             Message = request.env['mail.message'].sudo()
+
+        messages = Message.search(domain).filtered(
+            self._get_void_portal_messages_filter()
+        )
+
         return {
-            'messages': Message.search(domain, limit=limit, offset=offset).portal_message_format(),
-            'message_count': Message.search_count(domain)
+            "messages": messages[offset : offset + limit].portal_message_format(),
```

Comment:
Hello, post-filtering with offset/limit is never a working pattern.

Have only problematic messages on a page -> no result.

It must be built into the domain, or additional steps have to be taken to ensure "filtered messages" are replaced by other one, which gets very complicated.

Current solution of fetching all messages at the same time and applying limit/offset in python is not a solution that can scale. If there are a lot of messages on thread this will just be extremely slow, and eventually lead to out of memory errors.

It's should be easier to build it into the domain instead.